File: data-analystics/redis/code/chapter9/java/src/main/java/com/redis/tutorial/CachePatterns.java
package com.redis.tutorial;

import redis.clients.jedis.Jedis;
import redis.clients.jedis.params.SetParams;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Redis缓存模式实现
 * 包含缓存穿透、雪崩、击穿等问题的解决方案
 */
public class CachePatterns {
    
    private static final Logger logger = LoggerFactory.getLogger(CachePatterns.class);
    private static final ObjectMapper objectMapper = new ObjectMapper();
    
    private final Jedis jedis;
    private final String prefix;
    
    public CachePatterns(Jedis jedis, String prefix) {
        this.jedis = jedis;
        this.prefix = prefix;
    }
    
    public CachePatterns(Jedis jedis) {
        this(jedis, "cache:");
    }
    
    /**
     * 获取完整的缓存key
     */
    private String getKey(String key) {
        return prefix + key;
    }
    
    /**
     * 设置缓存
     */
    public boolean set(String key, Object value, int expireSeconds) {
        try {
            String cacheKey = getKey(key);
            String serializedValue = objectMapper.writeValueAsString(value);
            
            SetParams setParams = SetParams.setParams().ex(expireSeconds);
            String result = jedis.set(cacheKey, serializedValue, setParams);
            
            return "OK".equals(result);
        } catch (JsonProcessingException e) {
            logger.error("缓存序列化失败: {}", e.getMessage());
            return false;
        }
    }
    
    /**
     * 获取缓存
     */
    public <T> T get(String key, Class<T> clazz) {
        try {
            String cacheKey = getKey(key);
            String cachedData = jedis.get(cacheKey);
            
            if (cachedData == null) {
                return null;
            }
            
            return objectMapper.readValue(cachedData, clazz);
        } catch (Exception e) {
            logger.error("缓存反序列化失败: {}", e.getMessage());
            return null;
        }
    }
    
    /**
     * 删除缓存
     */
    public boolean delete(String key) {
        String cacheKey = getKey(key);
        Long result = jedis.del(cacheKey);
        return result != null && result > 0;
    }
}

/**
 * 布隆过滤器缓存（解决缓存穿透）
 */
class BloomFilterCache extends CachePatterns {
    
    private final String bloomPrefix;
    private final int bitSize;
    private final int hashCount;
    
    public BloomFilterCache(Jedis jedis, String prefix, String bloomPrefix, int bitSize, int hashCount) {
        super(jedis, prefix);
        this.bloomPrefix = bloomPrefix;
        this.bitSize = bitSize;
        this.hashCount = hashCount;
    }
    
    public BloomFilterCache(Jedis jedis) {
        this(jedis, "cache:", "bloom:", 1 << 20, 5);
    }
    
    /**
     * 计算布隆过滤器的多个hash值
     */
    private List<Integer> getHashValues(String key) {
        List<Integer> hashValues = new ArrayList<>();
        
        for (int i = 0; i < hashCount; i++) {
            // 使用不同的盐值计算hash
            String saltedKey = key + ":" + i;
            int hashVal = Math.abs(saltedKey.hashCode()) % bitSize;
            hashValues.add(hashVal);
        }
        
        return hashValues;
    }
    
    /**
     * 将key添加到布隆过滤器
     */
    public boolean addToBloom(String key) {
        String bloomKey = bloomPrefix + "filter";
        List<Integer> hashValues = getHashValues(key);
        
        // 使用管道批量设置位
        var pipeline = jedis.pipelined();
        for (int hashVal : hashValues) {
            pipeline.setbit(bloomKey, hashVal, true);
        }
        
        pipeline.sync();
        return true;
    }
    
    /**
     * 检查key是否可能在布隆过滤器中
     */
    public boolean existsInBloom(String key) {
        String bloomKey = bloomPrefix + "filter";
        List<Integer> hashValues = getHashValues(key);
        
        // 使用管道批量检查位
        var pipeline = jedis.pipelined();
        for (int hashVal : hashValues) {
            pipeline.getbit(bloomKey, hashVal);
        }
        
        List<Object> results = pipeline.syncAndReturnAll();
        
        // 所有位都为1才认为可能存在
        for (Object result : results) {
            if (!(result instanceof Boolean) || !(Boolean) result) {
                return false;
            }
        }
        
        return true;
    }
    
    /**
     * 带缓存穿透保护的获取方法
     */
    public <T> T getWithPenetrationProtection(String key, Class<T> clazz, DataFetcher<T> dataFetcher) {
        // 先检查布隆过滤器
        if (!existsInBloom(key)) {
            logger.info("Key {} 不在布隆过滤器中，直接返回空", key);
            return null;
        }
        
        // 检查缓存
        T cachedData = get(key, clazz);
        if (cachedData != null) {
            return cachedData;
        }
        
        // 从数据源获取数据
        T data = dataFetcher.fetch(key);
        
        if (data != null) {
            // 数据存在，设置缓存
            set(key, data, 3600);
            // 添加到布隆过滤器
            addToBloom(key);
        } else {
            // 数据不存在，设置空值缓存（短时间）
            set(key, "__NULL__", 300); // 5分钟
        }
        
        return data;
    }
}

/**
 * 缓存击穿保护（热点key失效保护）
 */
class CacheBreakdownProtection extends CachePatterns {
    
    private final Map<String, Lock> localLocks = new ConcurrentHashMap<>();
    private final Lock lockLock = new ReentrantLock();
    
    public CacheBreakdownProtection(Jedis jedis, String prefix) {
        super(jedis, prefix);
    }
    
    public CacheBreakdownProtection(Jedis jedis) {
        super(jedis);
    }
    
    /**
     * 获取或创建本地锁
     */
    private Lock getLocalLock(String key) {
        lockLock.lock();
        try {
            return localLocks.computeIfAbsent(key, k -> new ReentrantLock());
        } finally {
            lockLock.unlock();
        }
    }
    
    /**
     * 带缓存击穿保护的获取方法
     */
    public <T> T getWithBreakdownProtection(String key, Class<T> clazz, 
                                           DataFetcher<T> dataFetcher, int expireSeconds) {
        // 第一次尝试获取缓存
        T cachedData = get(key, clazz);
        if (cachedData != null) {
            return cachedData;
        }
        
        // 获取本地锁
        Lock localLock = getLocalLock(key);
        
        // 尝试获取分布式锁
        String lockKey = "lock:" + key;
        boolean lockAcquired = false;
        
        try {
            // 先尝试获取本地锁（避免不必要的分布式锁竞争）
            if (localLock.tryLock()) {
                try {
                    // 再次检查缓存（双重检查锁定）
                    cachedData = get(key, clazz);
                    if (cachedData != null) {
                        return cachedData;
                    }
                    
                    // 获取分布式锁
                    lockAcquired = acquireDistributedLock(lockKey, 10);
                    
                    if (lockAcquired) {
                        // 第三次检查缓存
                        cachedData = get(key, clazz);
                        if (cachedData != null) {
                            return cachedData;
                        }
                        
                        // 从数据源获取数据
                        T data = dataFetcher.fetch(key);
                        
                        if (data != null) {
                            // 设置缓存
                            set(key, data, expireSeconds);
                        }
                        
                        return data;
                    } else {
                        // 等待其他线程加载数据
                        for (int i = 0; i < 50; i++) { // 最多等待5秒
                            try {
                                TimeUnit.MILLISECONDS.sleep(100);
                            } catch (InterruptedException e) {
                                Thread.currentThread().interrupt();
                                break;
                            }
                            
                            cachedData = get(key, clazz);
                            if (cachedData != null) {
                                return cachedData;
                            }
                        }
                        
                        // 超时后直接查询数据源
                        return dataFetcher.fetch(key);
                    }
                } finally {
                    localLock.unlock();
                }
            } else {
                // 等待本地锁
                localLock.lock();
                try {
                    // 再次检查缓存
                    cachedData = get(key, clazz);
                    return cachedData != null ? cachedData : dataFetcher.fetch(key);
                } finally {
                    localLock.unlock();
                }
            }
        } finally {
            if (lockAcquired) {
                releaseDistributedLock(lockKey);
            }
            
            // 清理本地锁（可选）
            lockLock.lock();
            try {
                localLocks.remove(key);
            } finally {
                lockLock.unlock();
            }
        }
    }
    
    /**
     * 获取分布式锁
     */
    private boolean acquireDistributedLock(String lockKey, int timeoutSeconds) {
        long endTime = System.currentTimeMillis() + TimeUnit.SECONDS.toMillis(timeoutSeconds);
        String identifier = String.valueOf(System.currentTimeMillis());
        
        while (System.currentTimeMillis() < endTime) {
            SetParams setParams = SetParams.setParams().nx().ex(30);
            String result = jedis.set(lockKey, identifier, setParams);
            
            if ("OK".equals(result)) {
                return true;
            }
            
            try {
                TimeUnit.MILLISECONDS.sleep(100);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
        }
        
        return false;
    }
    
    /**
     * 释放分布式锁
     */
    private boolean releaseDistributedLock(String lockKey) {
        try {
            jedis.del(lockKey);
            return true;
        } catch (Exception e) {
            return false;
        }
    }
}

/**
 * 缓存雪崩保护（大量缓存同时失效保护）
 */
class CacheAvalancheProtection extends CachePatterns {
    
    public CacheAvalancheProtection(Jedis jedis, String prefix) {
        super(jedis, prefix);
    }
    
    public CacheAvalancheProtection(Jedis jedis) {
        super(jedis);
    }
    
    /**
     * 设置带雪崩保护的缓存
     */
    public boolean setWithAvalancheProtection(String key, Object value, 
                                             int baseExpireSeconds, int randomRangeSeconds) {
        // 添加随机偏移，避免同时失效
        int randomOffset = Math.abs(key.hashCode()) % randomRangeSeconds;
        int actualExpireSeconds = baseExpireSeconds + randomOffset;
        
        return set(key, value, actualExpireSeconds);
    }
    
    /**
     * 批量设置带雪崩保护的缓存
     */
    public boolean batchSetWithAvalancheProtection(Map<String, Object> dataMap, 
                                                  int baseExpireSeconds, int randomRangeSeconds) {
        try {
            var pipeline = jedis.pipelined();
            
            for (Map.Entry<String, Object> entry : dataMap.entrySet()) {
                String cacheKey = getKey(entry.getKey());
                int randomOffset = Math.abs(entry.getKey().hashCode()) % randomRangeSeconds;
                int actualExpireSeconds = baseExpireSeconds + randomOffset;
                
                try {
                    String serializedValue = objectMapper.writeValueAsString(entry.getValue());
                    pipeline.setex(cacheKey, actualExpireSeconds, serializedValue);
                } catch (JsonProcessingException e) {
                    logger.error("缓存序列化失败: {}", e.getMessage());
                }
            }
            
            pipeline.sync();
            return true;
        } catch (Exception e) {
            logger.error("批量设置缓存失败: {}", e.getMessage());
            return false;
        }
    }
}

/**
 * 多级缓存（本地缓存 + Redis缓存）
 */
class MultiLevelCache<T> {
    
    private final CachePatterns redisCache;
    private final Map<String, CacheEntry<T>> localCache = new ConcurrentHashMap<>();
    private final int localCacheSize;
    private final long localCacheTtl;
    private final Map<String, Long> accessTimes = new ConcurrentHashMap<>();
    
    public MultiLevelCache(Jedis jedis, int localCacheSize, long localCacheTtl) {
        this.redisCache = new CachePatterns(jedis);
        this.localCacheSize = localCacheSize;
        this.localCacheTtl = localCacheTtl;
    }
    
    /**
     * 从多级缓存中获取数据
     */
    public T get(String key, DataFetcher<T> dataFetcher) {
        long currentTime = System.currentTimeMillis();
        
        // 第一级：本地缓存
        CacheEntry<T> cacheEntry = localCache.get(key);
        if (cacheEntry != null) {
            if (currentTime - cacheEntry.timestamp < localCacheTtl) {
                logger.info("从本地缓存获取: {}", key);
                accessTimes.put(key, currentTime);
                return cacheEntry.data;
            } else {
                // 本地缓存过期
                localCache.remove(key);
                accessTimes.remove(key);
            }
        }
        
        // 第二级：Redis缓存
        @SuppressWarnings("unchecked")
        T cachedData = (T) redisCache.get(key, Object.class);
        if (cachedData != null) {
            logger.info("从Redis缓存获取: {}", key);
            // 更新本地缓存
            updateLocalCache(key, cachedData, currentTime);
            return cachedData;
        }
        
        // 第三级：数据库查询
        logger.info("缓存未命中，查询数据库: {}", key);
        T data = dataFetcher.fetch(key);
        
        if (data != null) {
            // 更新Redis缓存
            redisCache.set(key, data, 3600);
            // 更新本地缓存
            updateLocalCache(key, data, currentTime);
        }
        
        return data;
    }
    
    /**
     * 更新本地缓存
     */
    private void updateLocalCache(String key, T data, long timestamp) {
        // 检查缓存大小，如果超过限制则清理最久未使用的
        if (localCache.size() >= localCacheSize) {
            // 找到最久未使用的key
            String oldestKey = accessTimes.entrySet().stream()
                    .min(Map.Entry.comparingByValue())
                    .map(Map.Entry::getKey)
                    .orElse(null);
                    
            if (oldestKey != null) {
                localCache.remove(oldestKey);
                accessTimes.remove(oldestKey);
            }
        }
        
        localCache.put(key, new CacheEntry<>(data, timestamp));
        accessTimes.put(key, timestamp);
    }
    
    /**
     * 失效缓存
     */
    public void invalidate(String key) {
        localCache.remove(key);
        accessTimes.remove(key);
        redisCache.delete(key);
    }
    
    /**
     * 缓存条目
     */
    private static class CacheEntry<T> {
        final T data;
        final long timestamp;
        
        CacheEntry(T data, long timestamp) {
            this.data = data;
            this.timestamp = timestamp;
        }
    }
}

/**
 * 数据获取器接口
 */
@FunctionalInterface
interface DataFetcher<T> {
    T fetch(String key);
}

/**
 * 缓存模式测试类
 */
class CachePatternsTest {
    
    public static void main(String[] args) {
        Jedis jedis = RedisConfig.getJedis();
        
        try {
            System.out.println("=== 测试基础缓存 ===");
            CachePatterns basicCache = new CachePatterns(jedis);
            
            // 测试基础缓存操作
            Map<String, Object> testData = Map.of("name", "test", "value", 123);
            basicCache.set("test_key", testData, 3600);
            
            @SuppressWarnings("unchecked")
            Map<String, Object> result = (Map<String, Object>) basicCache.get("test_key", Object.class);
            System.out.println("基础缓存获取结果: " + result);
            
            System.out.println("\n=== 测试布隆过滤器缓存 ===");
            BloomFilterCache bloomCache = new BloomFilterCache(jedis);
            
            // 模拟数据源
            DataFetcher<String> mockDbFetcher = key -> {
                if (key.startsWith("exist_")) {
                    return "data_for_" + key;
                }
                return null;
            };
            
            // 测试存在的数据
            String result1 = bloomCache.getWithPenetrationProtection("exist_001", String.class, mockDbFetcher);
            System.out.println("存在数据查询结果: " + result1);
            
            // 测试不存在的数据
            String result2 = bloomCache.getWithPenetrationProtection("not_exist_001", String.class, mockDbFetcher);
            System.out.println("不存在数据查询结果: " + result2);
            
            System.out.println("\n=== 测试缓存雪崩保护 ===");
            CacheAvalancheProtection avalancheCache = new CacheAvalancheProtection(jedis);
            
            // 批量设置缓存
            Map<String, Object> testData2 = new HashMap<>();
            for (int i = 0; i < 10; i++) {
                testData2.put("key_" + i, "value_" + i);
            }
            avalancheCache.batchSetWithAvalancheProtection(testData2, 3600, 600);
            
            System.out.println("批量缓存设置完成");
            
            System.out.println("\n=== 测试多级缓存 ===");
            MultiLevelCache<String> multiCache = new MultiLevelCache<>(jedis, 5, TimeUnit.MINUTES.toMillis(5));
            
            DataFetcher<String> simpleFetcher = key -> "fetched_" + key;
            
            // 第一次访问
            String result3 = multiCache.get("test_key", simpleFetcher);
            System.out.println("第一次访问结果: " + result3);
            
            // 第二次访问（应该从本地缓存获取）
            String result4 = multiCache.get("test_key", simpleFetcher);
            System.out.println("第二次访问结果: " + result4);
            
            System.out.println("\n缓存模式测试完成!");
            
        } finally {
            RedisConfig.closeJedis(jedis);
        }
    }
}

File: data-analystics/redis/code/chapter9/python/cache_patterns.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Redis缓存模式实现
包含缓存穿透、雪崩、击穿等问题的解决方案
"""

import redis
import time
import json
import hashlib
from typing import Any, Optional, Callable
from threading import Lock


class RedisCache:
    """Redis缓存基础类"""
    
    def __init__(self, redis_client: redis.Redis, prefix: str = "cache:"):
        self.redis = redis_client
        self.prefix = prefix
    
    def get_key(self, key: str) -> str:
        """生成完整的缓存key"""
        return f"{self.prefix}{key}"
    
    def set(self, key: str, value: Any, expire: int = 3600) -> bool:
        """设置缓存"""
        try:
            cache_key = self.get_key(key)
            serialized_value = json.dumps(value)
            return self.redis.setex(cache_key, expire, serialized_value)
        except Exception as e:
            print(f"缓存设置失败: {e}")
            return False
    
    def get(self, key: str) -> Optional[Any]:
        """获取缓存"""
        try:
            cache_key = self.get_key(key)
            cached_data = self.redis.get(cache_key)
            
            if cached_data is None:
                return None
            
            return json.loads(cached_data)
        except Exception as e:
            print(f"缓存获取失败: {e}")
            return None
    
    def delete(self, key: str) -> bool:
        """删除缓存"""
        try:
            cache_key = self.get_key(key)
            return self.redis.delete(cache_key) > 0
        except Exception as e:
            print(f"缓存删除失败: {e}")
            return False


class BloomFilterCache(RedisCache):
    """布隆过滤器缓存（解决缓存穿透）"""
    
    def __init__(self, redis_client: redis.Redis, prefix: str = "cache:", 
                 bloom_prefix: str = "bloom:", bit_size: int = 2**20, hash_count: int = 5):
        super().__init__(redis_client, prefix)
        self.bloom_prefix = bloom_prefix
        self.bit_size = bit_size
        self.hash_count = hash_count
    
    def _get_hash_values(self, key: str) -> list:
        """计算布隆过滤器的多个hash值"""
        hash_values = []
        
        for i in range(self.hash_count):
            # 使用不同的盐值计算hash
            salted_key = f"{key}:{i}"
            hash_val = int(hashlib.md5(salted_key.encode()).hexdigest(), 16) % self.bit_size
            hash_values.append(hash_val)
        
        return hash_values
    
    def add_to_bloom(self, key: str) -> bool:
        """将key添加到布隆过滤器"""
        bloom_key = f"{self.bloom_prefix}filter"
        hash_values = self._get_hash_values(key)
        
        # 使用管道批量设置位
        pipeline = self.redis.pipeline()
        for hash_val in hash_values:
            pipeline.setbit(bloom_key, hash_val, 1)
        
        pipeline.execute()
        return True
    
    def exists_in_bloom(self, key: str) -> bool:
        """检查key是否可能在布隆过滤器中"""
        bloom_key = f"{self.bloom_prefix}filter"
        hash_values = self._get_hash_values(key)
        
        # 使用管道批量检查位
        pipeline = self.redis.pipeline()
        for hash_val in hash_values:
            pipeline.getbit(bloom_key, hash_val)
        
        results = pipeline.execute()
        
        # 所有位都为1才认为可能存在
        return all(results)
    
    def get_with_penetration_protection(self, key: str, data_fetcher: Callable) -> Optional[Any]:
        """带缓存穿透保护的获取方法"""
        # 先检查布隆过滤器
        if not self.exists_in_bloom(key):
            print(f"Key {key} 不在布隆过滤器中，直接返回空")
            return None
        
        # 检查缓存
        cached_data = self.get(key)
        if cached_data is not None:
            return cached_data
        
        # 从数据源获取数据
        data = data_fetcher(key)
        
        if data is not None:
            # 数据存在，设置缓存
            self.set(key, data)
            # 添加到布隆过滤器
            self.add_to_bloom(key)
        else:
            # 数据不存在，设置空值缓存（短时间）
            self.set(key, "__NULL__", expire=300)  # 5分钟
        
        return data


class CacheBreakdownProtection(RedisCache):
    """缓存击穿保护（热点key失效保护）"""
    
    def __init__(self, redis_client: redis.Redis, prefix: str = "cache:"):
        super().__init__(redis_client, prefix)
        self.local_locks = {}  # 本地锁字典
        self.lock_lock = Lock()  # 保护本地锁字典的锁
    
    def get_local_lock(self, key: str) -> Lock:
        """获取或创建本地锁"""
        with self.lock_lock:
            if key not in self.local_locks:
                self.local_locks[key] = Lock()
            return self.local_locks[key]
    
    def get_with_breakdown_protection(self, key: str, data_fetcher: Callable, 
                                    expire: int = 3600) -> Optional[Any]:
        """带缓存击穿保护的获取方法"""
        # 第一次尝试获取缓存
        cached_data = self.get(key)
        if cached_data is not None:
            return cached_data
        
        # 获取本地锁
        local_lock = self.get_local_lock(key)
        
        # 尝试获取分布式锁
        lock_key = f"lock:{key}"
        lock_acquired = False
        
        try:
            # 先尝试获取本地锁（避免不必要的分布式锁竞争）
            if local_lock.acquire(blocking=False):
                # 再次检查缓存（双重检查锁定）
                cached_data = self.get(key)
                if cached_data is not None:
                    return cached_data
                
                # 获取分布式锁
                lock_acquired = self._acquire_distributed_lock(lock_key)
                
                if lock_acquired:
                    # 第三次检查缓存
                    cached_data = self.get(key)
                    if cached_data is not None:
                        return cached_data
                    
                    # 从数据源获取数据
                    data = data_fetcher(key)
                    
                    if data is not None:
                        # 设置缓存
                        self.set(key, data, expire)
                    
                    return data
                else:
                    # 等待其他线程加载数据
                    for i in range(50):  # 最多等待5秒
                        time.sleep(0.1)
                        cached_data = self.get(key)
                        if cached_data is not None:
                            return cached_data
                    
                    # 超时后直接查询数据源
                    return data_fetcher(key)
            else:
                # 等待本地锁
                local_lock.acquire()
                try:
                    # 再次检查缓存
                    cached_data = self.get(key)
                    return cached_data if cached_data is not None else data_fetcher(key)
                finally:
                    local_lock.release()
        
        finally:
            if lock_acquired:
                self._release_distributed_lock(lock_key)
            
            # 清理本地锁（可选）
            with self.lock_lock:
                if key in self.local_locks:
                    try:
                        local_lock.release()
                        del self.local_locks[key]
                    except:
                        pass
    
    def _acquire_distributed_lock(self, lock_key: str, timeout: int = 10) -> bool:
        """获取分布式锁"""
        end_time = time.time() + timeout
        identifier = str(time.time())
        
        while time.time() < end_time:
            if self.redis.set(lock_key, identifier, ex=30, nx=True):
                return True
            time.sleep(0.1)
        
        return False
    
    def _release_distributed_lock(self, lock_key: str) -> bool:
        """释放分布式锁"""
        try:
            self.redis.delete(lock_key)
            return True
        except:
            return False


class CacheAvalancheProtection(RedisCache):
    """缓存雪崩保护（大量缓存同时失效保护）"""
    
    def __init__(self, redis_client: redis.Redis, prefix: str = "cache:"):
        super().__init__(redis_client, prefix)
    
    def set_with_avalanche_protection(self, key: str, value: Any, 
                                     base_expire: int = 3600, 
                                     random_range: int = 600) -> bool:
        """
        设置带雪崩保护的缓存
        
        Args:
            base_expire: 基础过期时间
            random_range: 随机范围（秒）
        """
        # 添加随机偏移，避免同时失效
        random_offset = hash(key) % random_range  # 使用key的hash确保一致性
        actual_expire = base_expire + random_offset
        
        return self.set(key, value, actual_expire)
    
    def batch_set_with_avalanche_protection(self, data_dict: dict, 
                                           base_expire: int = 3600, 
                                           random_range: int = 600) -> bool:
        """批量设置带雪崩保护的缓存"""
        try:
            pipeline = self.redis.pipeline()
            
            for key, value in data_dict.items():
                cache_key = self.get_key(key)
                random_offset = hash(key) % random_range
                actual_expire = base_expire + random_offset
                serialized_value = json.dumps(value)
                
                pipeline.setex(cache_key, actual_expire, serialized_value)
            
            pipeline.execute()
            return True
        except Exception as e:
            print(f"批量设置缓存失败: {e}")
            return False


class MultiLevelCache:
    """多级缓存（本地缓存 + Redis缓存）"""
    
    def __init__(self, redis_client: redis.Redis, local_cache_size: int = 1000):
        self.redis_cache = RedisCache(redis_client)
        self.local_cache = {}
        self.local_cache_size = local_cache_size
        self.local_cache_ttl = 300  # 本地缓存5分钟
        self.access_times = {}
    
    def get(self, key: str, data_fetcher: Callable) -> Optional[Any]:
        """从多级缓存中获取数据"""
        current_time = time.time()
        
        # 第一级：本地缓存
        if key in self.local_cache:
            timestamp, data = self.local_cache[key]
            if current_time - timestamp < self.local_cache_ttl:
                print(f"从本地缓存获取: {key}")
                return data
            else:
                # 本地缓存过期
                del self.local_cache[key]
                if key in self.access_times:
                    del self.access_times[key]
        
        # 第二级：Redis缓存
        cached_data = self.redis_cache.get(key)
        if cached_data is not None:
            print(f"从Redis缓存获取: {key}")
            # 更新本地缓存
            self._update_local_cache(key, cached_data, current_time)
            return cached_data
        
        # 第三级：数据库查询
        print(f"缓存未命中，查询数据库: {key}")
        data = data_fetcher(key)
        
        if data is not None:
            # 更新Redis缓存
            self.redis_cache.set(key, data)
            # 更新本地缓存
            self._update_local_cache(key, data, current_time)
        
        return data
    
    def _update_local_cache(self, key: str, data: Any, timestamp: float):
        """更新本地缓存"""
        # 检查缓存大小，如果超过限制则清理最久未使用的
        if len(self.local_cache) >= self.local_cache_size:
            # 找到最久未使用的key
            oldest_key = min(self.access_times.items(), key=lambda x: x[1])[0]
            del self.local_cache[oldest_key]
            del self.access_times[oldest_key]
        
        self.local_cache[key] = (timestamp, data)
        self.access_times[key] = timestamp
    
    def invalidate(self, key: str):
        """失效缓存"""
        if key in self.local_cache:
            del self.local_cache[key]
        if key in self.access_times:
            del self.access_times[key]
        
        self.redis_cache.delete(key)


def test_cache_patterns():
    """测试缓存模式"""
    # 创建Redis连接
    r = redis.Redis(host='localhost', port=6379, db=0, decode_responses=True)
    
    print("=== 测试基础缓存 ===")
    basic_cache = RedisCache(r)
    
    # 测试基础缓存操作
    basic_cache.set("test_key", {"name": "test", "value": 123})
    result = basic_cache.get("test_key")
    print(f"基础缓存获取结果: {result}")
    
    print("\n=== 测试布隆过滤器缓存 ===")
    bloom_cache = BloomFilterCache(r)
    
    # 模拟数据源
    def mock_db_fetcher(key):
        if key.startswith("exist_"):
            return f"data_for_{key}"
        return None
    
    # 测试存在的数据
    result1 = bloom_cache.get_with_penetration_protection("exist_001", mock_db_fetcher)
    print(f"存在数据查询结果: {result1}")
    
    # 测试不存在的数据
    result2 = bloom_cache.get_with_penetration_protection("not_exist_001", mock_db_fetcher)
    print(f"不存在数据查询结果: {result2}")
    
    print("\n=== 测试缓存击穿保护 ===")
    breakdown_cache = CacheBreakdownProtection(r)
    
    call_count = 0
    def expensive_fetcher(key):
        nonlocal call_count
        call_count += 1
        print(f"数据源被调用第 {call_count} 次")
        time.sleep(0.5)  # 模拟耗时操作
        return f"expensive_data_for_{key}"
    
    # 模拟并发访问
    import threading
    
    results = []
    
    def concurrent_get():
        result = breakdown_cache.get_with_breakdown_protection("hot_key", expensive_fetcher)
        results.append(result)
    
    threads = []
    for i in range(10):
        t = threading.Thread(target=concurrent_get)
        threads.append(t)
        t.start()
    
    for t in threads:
        t.join()
    
    print(f"并发访问结果数量: {len(results)}")
    print(f"数据源实际调用次数: {call_count}")
    
    print("\n=== 测试缓存雪崩保护 ===")
    avalanche_cache = CacheAvalancheProtection(r)
    
    # 批量设置缓存
    test_data = {f"key_{i}": f"value_{i}" for i in range(10)}
    avalanche_cache.batch_set_with_avalanche_protection(test_data)
    
    print("批量缓存设置完成")
    
    print("\n=== 测试多级缓存 ===")
    multi_cache = MultiLevelCache(r, local_cache_size=5)
    
    def simple_fetcher(key):
        return f"fetched_{key}"
    
    # 第一次访问
    result1 = multi_cache.get("test_key", simple_fetcher)
    print(f"第一次访问结果: {result1}")
    
    # 第二次访问（应该从本地缓存获取）
    result2 = multi_cache.get("test_key", simple_fetcher)
    print(f"第二次访问结果: {result2}")
    
    print("\n缓存模式测试完成!")


if __name__ == "__main__":
    test_cache_patterns()